### PR TITLE
[Fix] Fix cuda graph padding for triton attention backend

### PR DESCRIPTION
**Pull request by merrymercy**
- We should initialize `req_to_token` with `torch.zeros` instead of `torch.empty` to avoid out-of-bound access during cuda graph capture.
- Then we do not need `get_cuda_graph_seq_len_fill_value`. We can always use 1.

### CODE REVIEW

**Comment by zhyncs** [comment]
I tried this branch with `python3 -m sglang.launch_server --model-path meta-llama/Llama-3.1-70B-Instruct --host 0.0.0.0 --port 30000 --disable-radix-cache --tp 4 --enable-overlap-schedule` on H100 and it seems doesn't work for me. BTW I use the latest docker image and this branch.

```
Overlap scheduler mode is enabled. This is an experimental feature. Sampling penalizer (e.g., frequency and repetition penalty), constrained decoding (e.g., regex, JSON), and embedding APIs are not supported and will lead to wrong results. The NaN detection is also disabled.
[2024-10-24 11:51:53] server_args=ServerArgs(model_path='meta-llama/Llama-3.1-70B-Instruct', tokenizer_path='meta-llama/Llama-3.1-70B-Instruct', tokenizer_mode='auto', skip_tokenizer_init=False, load_format='auto', trust_remote_code=False, dtype='auto', kv_cache_dtype='auto', quantization=None, context_length=None, device='cuda', served_model_name='meta-llama/Llama-3.1-70B-Instruct', chat_template=None, is_embedding=False, host='0.0.0.0', port=30000, mem_fraction_static=0.85, max_running_requests=None, max_total_tokens=None, chunked_prefill_size=8192, max_prefill_tokens=16384, schedule_policy='lpm', schedule_conservativeness=1.0, tp_size=4, stream_interval=1, random_seed=139307397, constrained_json_whitespace_pattern=None, log_level='info', log_level_http=None, log_requests=False, show_time_cost=False, api_key=None, file_storage_pth='SGLang_storage', enable_cache_report=False, dp_size=1, load_balance_method='round_robin', dist_init_addr=None, nnodes=1, node_rank=0, json_model_override_args='{}', enable_double_sparsity=False, ds_channel_config_path=None, ds_heavy_channel_num=32, ds_heavy_token_num=256, ds_heavy_channel_type='qk', ds_sparse_decode_threshold=4096, lora_paths=None, max_loras_per_batch=8, attention_backend='flashinfer', sampling_backend='flashinfer', disable_flashinfer=False, disable_flashinfer_sampling=False, disable_radix_cache=True, disable_regex_jump_forward=False, disable_cuda_graph=False, disable_cuda_graph_padding=False, disable_disk_cache=False, disable_custom_all_reduce=False, disable_mla=False, disable_penalizer=True, disable_nan_detection=True, enable_overlap_schedule=True, enable_mixed_chunk=False, enable_torch_compile=False, max_torch_compile_bs=32, torchao_config='', enable_p2p_check=False, triton_attention_reduce_in_fp32=False, num_continuous_decode_steps=1)
[2024-10-24 11:52:05 TP2] Init torch distributed begin.
[2024-10-24 11:52:05 TP1] Init torch distributed begin.
[2024-10-24 11:52:05 TP0] Init torch distributed begin.
[2024-10-24 11:52:06 TP3] Init torch distributed begin.
[2024-10-24 11:52:14 TP0] Load weight begin. avail mem=76.83 GB
[2024-10-24 11:52:14 TP1] Load weight begin. avail mem=76.73 GB
[2024-10-24 11:52:14 TP3] Load weight begin. avail mem=77.20 GB
[2024-10-24 11:52:14 TP2] Load weight begin. avail mem=76.73 GB
[2024-10-24 11:52:14 TP1] Ignore import error when loading sglang.srt.models.mllama. No module named 'transformers.models.mllama'
[2024-10-24 11:52:14 TP2] Ignore import error when loading sglang.srt.models.mllama. No module named 'transformers.models.mllama'
[2024-10-24 11:52:14 TP3] Ignore import error when loading sglang.srt.models.mllama. No module named 'transformers.models.mllama'
[2024-10-24 11:52:14 TP1] Ignore import error when loading sglang.srt.models.qwen2_vl. No module named 'einops'
[2024-10-24 11:52:14 TP2] Ignore import error when loading sglang.srt.models.qwen2_vl. No module named 'einops'
[2024-10-24 11:52:14 TP3] Ignore import error when loading sglang.srt.models.qwen2_vl. No module named 'einops'
[2024-10-24 11:52:14 TP0] Ignore import error when loading sglang.srt.models.mllama. No module named 'transformers.models.mllama'
[2024-10-24 11:52:14 TP0] Ignore import error when loading sglang.srt.models.qwen2_vl. No module named 'einops'
INFO 10-24 11:52:14 weight_utils.py:236] Using model weights format ['*.safetensors']
INFO 10-24 11:52:14 weight_utils.py:236] Using model weights format ['*.safetensors']
INFO 10-24 11:52:15 weight_utils.py:236] Using model weights format ['*.safetensors']
INFO 10-24 11:52:15 weight_utils.py:236] Using model weights format ['*.safetensors']
Loading safetensors checkpoint shards:   0% Completed | 0/30 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:   3% Completed | 1/30 [00:00<00:12,  2.26it/s]
Loading safetensors checkpoint shards:   7% Completed | 2/30 [00:00<00:13,  2.03it/s]
Loading safetensors checkpoint shards:  10% Completed | 3/30 [00:01<00:13,  1.97it/s]
Loading safetensors checkpoint shards:  13% Completed | 4/30 [00:02<00:13,  1.94it/s]
Loading safetensors checkpoint shards:  17% Completed | 5/30 [00:02<00:13,  1.90it/s]
Loading safetensors checkpoint shards:  20% Completed | 6/30 [00:03<00:12,  1.89it/s]
Loading safetensors checkpoint shards:  23% Completed | 7/30 [00:03<00:11,  1.93it/s]
Loading safetensors checkpoint shards:  27% Completed | 8/30 [00:04<00:11,  1.94it/s]
Loading safetensors checkpoint shards:  30% Completed | 9/30 [00:04<00:10,  1.94it/s]
Loading safetensors checkpoint shards:  33% Completed | 10/30 [00:05<00:09,  2.02it/s]
Loading safetensors checkpoint shards:  37% Completed | 11/30 [00:05<00:09,  2.01it/s]
Loading safetensors checkpoint shards:  40% Completed | 12/30 [00:06<00:08,  2.00it/s]
Loading safetensors checkpoint shards:  43% Completed | 13/30 [00:06<00:08,  1.99it/s]
Loading safetensors checkpoint shards:  47% Completed | 14/30 [00:07<00:07,  2.01it/s]
Loading safetensors checkpoint shards:  50% Completed | 15/30 [00:07<00:07,  1.99it/s]
Loading safetensors checkpoint shards:  53% Completed | 16/30 [00:08<00:07,  1.96it/s]
Loading safetensors checkpoint shards:  57% Completed | 17/30 [00:08<00:06,  2.09it/s]
Loading safetensors checkpoint shards:  60% Completed | 18/30 [00:09<00:06,  1.99it/s]
Loading safetensors checkpoint shards:  63% Completed | 19/30 [00:09<00:05,  1.96it/s]
Loading safetensors checkpoint shards:  67% Completed | 20/30 [00:10<00:04,  2.02it/s]
Loading safetensors checkpoint shards:  70% Completed | 21/30 [00:10<00:04,  1.96it/s]
Loading safetensors checkpoint shards:  73% Completed | 22/30 [00:11<00:04,  1.94it/s]
Loading safetensors checkpoint shards:  77% Completed | 23/30 [00:11<00:03,  1.91it/s]
Loading safetensors checkpoint shards:  80% Completed | 24/30 [00:12<00:03,  1.87it/s]
Loading safetensors checkpoint shards:  83% Completed | 25/30 [00:12<00:02,  1.87it/s]
Loading safetensors checkpoint shards:  87% Completed | 26/30 [00:13<00:02,  1.93it/s]
Loading safetensors checkpoint shards:  90% Completed | 27/30 [00:13<00:01,  2.40it/s]
Loading safetensors checkpoint shards:  93% Completed | 28/30 [00:13<00:00,  2.28it/s]
Loading safetensors checkpoint shards:  97% Completed | 29/30 [00:14<00:00,  2.24it/s]
Loading safetensors checkpoint shards: 100% Completed | 30/30 [00:14<00:00,  2.24it/s]
Loading safetensors checkpoint shards: 100% Completed | 30/30 [00:14<00:00,  2.02it/s]

[2024-10-24 11:52:30 TP2] Load weight end. type=LlamaForCausalLM, dtype=torch.bfloat16, avail mem=43.82 GB
[2024-10-24 11:52:30 TP1] Load weight end. type=LlamaForCausalLM, dtype=torch.bfloat16, avail mem=43.82 GB
[2024-10-24 11:52:30 TP3] Load weight end. type=LlamaForCausalLM, dtype=torch.bfloat16, avail mem=44.29 GB
[2024-10-24 11:52:30 TP0] Load weight end. type=LlamaForCausalLM, dtype=torch.bfloat16, avail mem=43.91 GB
[2024-10-24 11:52:30 TP1] Memory pool end. avail mem=10.32 GB
[2024-10-24 11:52:30 TP2] Memory pool end. avail mem=10.32 GB
[2024-10-24 11:52:30 TP3] Memory pool end. avail mem=10.79 GB
[2024-10-24 11:52:30 TP0] Memory pool end. avail mem=10.41 GB
[2024-10-24 11:52:30 TP1] Capture cuda graph begin. This can take up to several minutes.
[2024-10-24 11:52:30 TP2] Capture cuda graph begin. This can take up to several minutes.
[2024-10-24 11:52:30 TP0] Capture cuda graph begin. This can take up to several minutes.
[2024-10-24 11:52:30 TP3] Capture cuda graph begin. This can take up to several minutes.
```

**Comment by zhyncs** [comment]
After I fixed some import errors in the log, it still got stuck on CUDA Graph.

**Comment by merrymercy** [comment]
@zhyncs Why are you interested in trying this branch? The CI seems good. I can revert the change of `get_cuda_graph_seq_len_fill_value` to fix your problem.

**Comment by zhyncs** [comment]
> @zhyncs Why are you interested in trying this branch? The CI seems good. I can revert the change of `get_cuda_graph_seq_len_fill_value` to fix your problem.

Because main branch also doesn't work

**Comment by merrymercy** [comment]
@zhyncs's problem is due to other things unrelated to this PR. We will debug offline

**Comment by HaiShaw** [comment]
@merrymercy , nice catch! This should fix the `GPU Memory Access Fault` on MI300x as well!
We had seen unexpected errors from `torch.empty` used in initialization in other projects, it does not do real `zeroisation`.
Do we need to change more places to `torch.zeros`?

**Comment by HaiShaw** [comment]
@merrymercy Memory Fault on AMD is gone. THANKS! 💯

**Comment by glen-amd** [comment]
> * We should initialize `req_to_token` with `torch.zeros` instead of `torch.empty` to avoid out-of-bound access during cuda graph capture.
> * Then we do not need `get_cuda_graph_seq_len_fill_value`. We can always use 1.

@merrymercy - a couple of confusions.

1) How did you spot the OOB access fault?
2) Can you please elaborate why `torch.empty()` may cause this kind of faults?
    Is it due to the random values in the uninitialized memory space?
    Did we by chance use these random values before we explicitly filled the space with meaningful values?
   If that's the case, can you please pointer me to the related code using these random values?
3) In a sense, `torch.empty()` vs `torch.zeros()` is like `malloc()` vs `calloc()`.
    From this PoV, I didn't figure out why OOB access happened.

Thanks.

**Comment by glen-amd** [comment]
> @merrymercy , nice catch! This should fix the `GPU Memory Access Fault` on MI300x as well! We had seen unexpected errors from `torch.empty` used in initialization in other projects, it does not do real `zeroisation`. Do we need to change more places to `torch.zeros`?

`torch.empty()` never does "zeroisation".
Even with some preconditions like `torch.utils.deterministic.fill_uninitialized_memory` and `torch.use_deterministic_algorithms()`, the enforced initialization values are NaN or `INT_MAX`.

**Comment by glen-amd** [comment]
Moreover, just like `malloc()`, a best practice for using `torch.empty()` is to explicitly fill values right after it returns.
E.g., https://github.com/pytorch/pytorch/blob/fe18a221eb35e603002b094ad8ceb353e2cdca56/benchmarks/gpt_fast/generate.py#L132-L133
```
empty = torch.empty(T_new, dtype=dtype, device=device)
empty[:T] = prompt
```
